### ISSUE TITLE
[Feature Request] Sticky/Pinned Threads

### ISSUE THREAD

**Issue by jyapayne** [nim-lang/nimforum#170]
The ability to mark threads as sticky/pinned and view them is highly useful for both moderators and users.

In order for users to get value out of a forum, there needs to be some way of consistently accessing reference material so that new users can learn about common issues/features/tutorials and also so that more experienced users can generate helpful content for new users easily, without needing to create a blog (or website), or file a pull request to make a guest post.

**Logistics**

A stickied post can be implemented by adding a field [here](https://github.com/nim-lang/nimforum/blob/master/src/frontend/threadlist.nim#L17) and adding the appropriate behavior on the front-end. A button for moderators must also be added in order to mark a post as stickied. A stickied post could also just be an ordinary tag/category (#169).

For the behavior on the front-end, there is one of several ways to implement it.

The first (and easiest) is to simply gather all stickied posts and put them at the top of the current post list. However, this would get crowded quickly and (left unsorted) could get just as crowded as the current post list. A checkbox could be added in order to view stickied posts.

The second (which requires categories #169), is to have them in a different view separate from the current post list, and sortable/filterable (#168) by category/tag. I'm in favor of this way, since then stickied posts would be more useful than just another list of unsorted (but curated) posts.

**Comment by zetashift** [comment]
I'm trying to work on this, I have added an `isSticky` field, and would to go the second route to implement the logic. So next to 'Latest' & 'Categories' there will be a 'Stickies' list and it be a view similiar to categories?

**Comment by dom96** [comment]
Wouldn't that mean that stickied threads aren't visible to people visiting the front page?

**Comment by jyapayne** [comment]
@zetashift Actually, looking back at this now, I think the first easiest option my be a good first pass. If they need to be in a separate view later, then they can be. No need to complicate things off the bat. Plus, if the admin controls for making a post stickied are implemented, then the admins can get rid of stickied posts if there are too many. What are your thoughts @dom96?

**Comment by zetashift** [comment]
Sorry for the late reaction, first option sounds good too. 
I am very new to the codebase and karax, so I'll be asking lots of question, bare with me please!

For the logic of actually showing the stickies first, I'm guessing I have to implement that here with something like an `if` statement: https://github.com/nim-lang/nimforum/blob/master/src/frontend/threadlist.nim#L220 any tips how I would do this efficiently?

**Comment by dom96** [comment]
All you should really need is to modify the `Thread` type, add `isSticky` to it and make sure it's written/read from the db correctly.

In the SQL that gets the thread list just add something like `OR isSticky == 'true'`. Then you might be able to order by stickiness to make them show up at the top in SQL or if not just do it in the Nim code. 

Once you've got that you just need to give them some different styles when rendering in the front end.

Then you can implement a way to make a thread sticky, probably just a button for the first post in a thread is easiest to implement (beside the heart button).

**Comment by zetashift** [comment]
Thank you, that makes more sense yeah, I've modified the SQL that inserts thread with an `isSticky` so far. For the actual query for getting the threads, I think that's happening here: https://github.com/nim-lang/nimforum/blob/master/src/forum.nim#L835 , am I right?

Can I order this query like this, so instead of 
`order by modified desc limit ?, ?;"""` at line 849, I do:
`order by isSticky, modified desc limit ?, ?;"""`
I'll test things out tomorrow, just tryin' to make sure I'm not touching stuff that doesn't have anything to do with getting threads haha.

**Comment by dom96** [comment]
It looks like you're on the right track :)

**Comment by timotheecour** [comment]
I don't understand the need for this feature TBH, all we need is better support for adding/editing labels to threads, along with support for labels in search; implementing a (subset of) what github offers in terms of labels (editing labels and searching with labels) would be IMO much more valuable instead of hardcoding something like sticky/pinned threads. eg:
```
is:issue is:open label:"OS/Arch specific" sort:updated-desc involves:dom96 
```

https://github.com/nim-lang/Nim/issues?q=is%3Aissue+is%3Aopen+label%3A%22OS%2FArch+specific%22+sort%3Aupdated-desc+involves%3Adom96

labels provide the most flexible UI.

**Comment by zetashift** [comment]
It's mostly a nice feature for discoverability, if there are 1/2 stickies in at the front it draws eyes to the first post immediately.
There is a reason why most forums have stickies, and if nimforum supports it, it might make it easier for end users too.

**Comment by dom96** [comment]
There are often cases when we may want to prioritise a thread, some examples off the top of my head:

* We released an RC for Nim, please test it!
* Join our weekly meeting, it's tomorrow!
* NimConf 2021!

**Comment by jyapayne** [comment]
Closed via #278
